### ISSUE TITLE
Trouble completing macro arguments

### ISSUE THREAD

**Issue by ckandeler** [clangd/clangd#1159]
Consider this piece of code:
```
#define MY_VARIADIC_MACRO(...)
#define MY_NORMAL_MACRO1(x)
#define MY_NORMAL_MACRO2(x) (x)

int main()
{
    int foo = 0;
    MY_VARIADIC_MACRO(f)   // No completion results for f
    MY_NORMAL_MACRO1(f)  // No completion results for f
    MY_NORMAL_MACRO2(f)  // f gets completed to foo, as expected
}

```
See inline comments for problem description.

Output of `clangd --version`:
clangd version 15.0.0 (https://github.com/llvm/llvm-project.git f707e1255e2f0a46c7a23271f594a9a4e5ec8f08)
Features: linux
Platform: x86_64-unknown-linux-gnu

**Comment by HighCommander4** [comment]
Clangd uses the macro expansion to determine the context in which to perform completion.

Consider:

```
#include <vector>

int globalVar;

struct MyClass {
  void foo();
  void bar();
};

#define STD(x) std::x
#define MYCLASS(x) MyClass().x
#define GLOBAL(x) x

int main() {
  STD(v);      // offers std::vector
  MYCLASS(f);  // offers MyClass::foo
  GLOBAL(g);   // offers globalVar
}
```

If the macro has no expansion, clangd doesn't know the context and doesn't offer any completions.

Editors usually have some sort of non-semantic fallback (e.g. vscode's `editor.wordBasedSuggestions`) that will offer all the words from the document, which is usually no worse than clangd could do without a completion context.

**Comment by HighCommander4** [comment]
> Editors usually have some sort of non-semantic fallback (e.g. vscode's `editor.wordBasedSuggestions`) that will offer all the words from the document, which is usually no worse than clangd could do without a completion context.

Technically, I suppose clangd could do _a bit_ better than the client by offering identifiers but not words that appear in comments/strings.

**Comment by HighCommander4** [comment]
I suppose clangd could also heuristically choose a context as a fallback, e.g. the macro invocation context (i.e. offer the same completions at `MACRO(x^` as if just `x^` had been written).

**Comment by Pesa** [comment]
> I suppose clangd could also heuristically choose a context as a fallback, e.g. the macro invocation context (i.e. offer the same completions at `MACRO(x^` as if just `x^` had been written).

That would be a huge improvement in my case, where the "non-working" macro is `BOOST_TEST` and normally clangd is perfectly capable of completing the expression if I type it outside the macro.

**Comment by HighCommander4** [comment]
> That would be a huge improvement in my case, where the "non-working" macro is `BOOST_TEST` and normally clangd is perfectly capable of completing the expression if I type it outside the macro.

I think it's worth filing a separate issue for this case. You're right that the heuristic described above would make it work, but `BOOST_TEST` is a macro whose argument _should_ appear in the expansion and provide a valid context, so the fact that it doesn't sounds like a separate bug.

**Comment by Hyderman** [comment]
I don't know if it's really related but there isn't any suggestion when defining macros.
``` cpp
#define AAAAAA (0)
#define BBBBBB (AAAAAA) // no suggestion for AAAAAA
```

**Comment by HighCommander4** [comment]
> I don't know if it's really related but there isn't any suggestion when defining macros.
> 
> ```c++
> #define AAAAAA (0)
> #define BBBBBB (AAAAAA) // no suggestion for AAAAAA
> ```

Yes, I think this is a variant of the same issue. In a macro definition, we have no information about what context the macro will be expanded in, but we could use the context of the macro definition as a heuristic (and so e.g. in this case offer completions from the global scope, such as previously defined macros).

**Comment by ckandeler** [comment]
> > That would be a huge improvement in my case, where the "non-working" macro is `BOOST_TEST` and normally clangd is perfectly capable of completing the expression if I type it outside the macro.
> 
> I think it's worth filing a separate issue for this case. You're right that the heuristic described above would make it work, but `BOOST_TEST` is a macro whose argument _should_ appear in the expansion and provide a valid context, so the fact that it doesn't sounds like a separate bug.

It seems it's simply a more round-about variant of the "non-referenced parameter" problem. I had AI distill a stand-alone reproducer from the boost macros:

```
#define TOOL(level, P) (P)
#define COUNT_(e0, e1, size, ...) size
#define COUNT(...) COUNT_(__VA_ARGS__, 2, 1)
#define IIF_1(t, f) f
#define IIF_2(t, f) t
#define IIF_I(cond, t, f) IIF_##cond(t, f)
#define IIF(cond, t, f) IIF_I(cond, t, f)
#define INVOKE(tool, ...) tool(__VA_ARGS__)
#define DISPATCH(...) INVOKE(IIF(COUNT(__VA_ARGS__), TOOL, TOOL), __VA_ARGS__)

void test() {
  int foobar = 42;
  DISPATCH(CHECK, foobar == foobar);  // completing "fooba|r": 0 results
}
```
According to Claude Code:
"It isolates precisely the two structural ingredients needed — (1) a variadic argument-counting macro that binds the completion-containing tokens to a named parameter its body doesn't use, purely to compute an arity, and (2) a separate named "invoke" macro that forwards the dispatch result through one more rescan."
